### PR TITLE
[SYCL-MLIR] Fix `memref2pointer(subindex)` optimization

### PR DESCRIPTION
**Pull request by whitneywhtsang**
Optimization `Memref2PointerIndex` is only valid when the element type of the source of subindex is not a struct type.
After this PR, `stream-copy.cpp` and `stream-triad.cpp` can both run successfully. They are added in https://github.com/intel/llvm-test-suite/pull/1384.

Signed-off-by: Tsang, Whitney <whitney.tsang@intel.com>

### CODE REVIEW

File: polygeist/test/polygeist-opt/invalid_canonicalization.mlir [hunk 1]
```diff
@@ -46,3 +46,19 @@ func.func @SimplifySubIndexUsers(%arg0: memref<?x!llvm.struct<(i32)>>) -> memref
 }
 
 // -----
+
+// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32> {
+// CHECK-NEXT:    [[C1:%.*]] = arith.constant 1 : index
+// CHECK-NEXT:    [[T0:%.*]] = "polygeist.subindex"([[A0]], [[C1]]) : (memref<?x!llvm.struct<(i32, i32)>>, index) -> memref<?xi32>
+// CHECK-NEXT:    [[T1:%.*]] = "polygeist.memref2pointer"([[T0]]) : (memref<?xi32>) -> !llvm.ptr<i32>
+// CHECK-NEXT:    return [[T1]] : !llvm.ptr<i32>
+// CHECK-NEXT:  }
```
Comment (victor-eds) [inline polygeist/test/polygeist-opt/invalid_canonicalization.mlir:55]: I'd argue that we'd ideally want something like:
```mlir
// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32> {
// CHECK-NEXT:    [[T0:%.*]] = "polygeist.memref2pointer"([[A0]]) : (memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<?x!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T1:%.*]] = llvm.getelementptr [[T0]][1, 0] : (!llvm.ptr<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32>
// CHECK-NEXT:    return [[T1]] : !llvm.ptr<i32>
// CHECK-NEXT:  }
```
If the types differed, we'd need to do something like:
```mlir
// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i8> {
// CHECK-NEXT:    [[T0:%.*]] = "polygeist.memref2pointer"([[A0]]) : (memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<?x!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T1:%.*]] = llvm.getelementptr [[T0]][1] : (!llvm.ptr<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<?x!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T2:%.*]] = llvm.bitcast([[T1]]) : !llvm.ptr<?x!llvm.struct<(i32, i32)>> to !llvm.ptr<i8>
// CHECK-NEXT:    return [[T2]] : !llvm.ptr<i8>
// CHECK-NEXT:  }
```
Can we address this in this PR?

**Review by victor-eds — CHANGES_REQUESTED** [review]

File: polygeist/test/polygeist-opt/invalid_canonicalization.mlir (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/invalid_canonicalization.mlir:55]: Given that `polygeist.subindex` is lowered to `llvm.getelementptr` in `ConvertPolygeistToLLVM` pass, why is it better to duplicate the logic and do this canonicalization here?

Assuming we want to minimize `polygeist.subindex`:
Should the first case be:
```
// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32> {
// CHECK-NEXT:    [[T0:%.*]] = "polygeist.memref2pointer"([[A0]]) : (memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T1:%.*]] = llvm.getelementptr [[T0]][0, 1] : (!llvm.ptr<!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32>
// CHECK-NEXT:    return [[T1]] : !llvm.ptr<i32>
// CHECK-NEXT:  }
```
And the second case be:
```
// CHECK:  func.func @Memref2PointerIndex([[A0:%.*]]: memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i8> {
// CHECK-NEXT:    [[T0:%.*]] = "polygeist.memref2pointer"([[A0]]) : (memref<?x!llvm.struct<(i32, i32)>>) -> !llvm.ptr<!llvm.struct<(i32, i32)>>
// CHECK-NEXT:    [[T1:%.*]] = llvm.getelementptr [[T0]][0, 1] : (!llvm.ptr<!llvm.struct<(i32, i32)>>) -> !llvm.ptr<i32>>
// CHECK-NEXT:    [[T2:%.*]] = llvm.bitcast([[T1]]) : !llvm.ptr<i32> to !llvm.ptr<i8>
// CHECK-NEXT:    return [[T2]] : !llvm.ptr<i8>
// CHECK-NEXT:  }
```
?

Comment (victor-eds) [reply polygeist/test/polygeist-opt/invalid_canonicalization.mlir:55]: Giving this a second though, I don't really see the point of having this canonicalization pattern in the first place. As you say, the lowering should be taking care of this. Do you think we should disable it altogether?

Comment (etiotto) [reply polygeist/test/polygeist-opt/invalid_canonicalization.mlir:55]: The transformation that this PR address is `Memref2PointerIndex`:
```
/// Simplify pointer2memref(memref2pointer(x)) to cast(x)
class Memref2PointerIndex final : public OpRewritePattern<Memref2PointerOp> {
```
According to the comment this transformation should 'fold' cack to pointer2memref, memref2pointer operations into a conversion. The new test case does not have that pattern so I'm surprised it consider this test case a candidate at all. 
Is the transformation actually doing something else (the comment is misleading)?

Comment (victor-eds) [reply polygeist/test/polygeist-opt/invalid_canonicalization.mlir:55]: The comment is misleading, I forgot to point it out in a comment.

Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/invalid_canonicalization.mlir:55]: Right, it should be more like
/// Simplify memref2pointer(subindex(x)) to getelementptr(memref2pointer(x))

Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/invalid_canonicalization.mlir:55]: > Giving this a second though, I don't really see the point of having this canonicalization pattern in the first place. As you say, the lowering should be taking care of this. Do you think we should disable it altogether?

Right, I don't see the point too, removed. 😄

**Review by victor-eds — APPROVED** [review]
LGTM

File: polygeist/test/polygeist-opt/invalid_canonicalization.mlir (hunk 1 above)
Comment (whitneywhtsang) [reply polygeist/test/polygeist-opt/invalid_canonicalization.mlir:55]: hmm..ok, if everyone is ok, I will revert the last commit and do it in a separate PR.

Comment (etiotto) [reply polygeist/test/polygeist-opt/invalid_canonicalization.mlir:55]: I'm OK with the fix.  I do think that this transformation should generate a `"polygeist.memref2pointer` with a result type that is a llvm.ptr<elemTy> where elemTy is identical to the original memref element type. If there is a need for a bitcast operation I think it should be injected after the GEP operation. That is not what this pass does, but in this PR I am happy we just fix the wrong codegen and leave the remaining consideration to the future (we might want to revisit all these polygeist canonicalization passes to make sure they make sense for our use cases).

**Review by etiotto — APPROVED** [review]

**Comment by victor-eds** [comment]
This leads me to the following topic: should we introduce a `memrefdiff` operation with the following lowering?:
```mlir
// 
memrefdiff(a, b) ->
c = memref2ptr(a)
d = memref2ptr(b)
e = ptr2int(c)
f = ptr2int(d)
diff = sub(e, f)
size = constant (sizeof element_ty)
res = div(diff, size)
```
I'd say canonicalizing and optimizing on this might be easier than on its lowering (which is our current representation)

**Comment by etiotto** [comment]
> This leads me to the following topic: should we introduce a `memrefdiff` operation with the following lowering?:
> 
> ```mlir
> // 
> memrefdiff(a, b) ->
> c = memref2ptr(a)
> d = memref2ptr(b)
> e = ptr2int(c)
> f = ptr2int(d)
> diff = sub(e, f)
> size = constant (sizeof element_ty)
> res = div(diff, size)
> ```
> 
> I'd say canonicalizing and optimizing on this might be easier than on its lowering (which is our current representation)

If I understand your comment correctly you are proposing to add a new operation to the memref dialect. The new operation would take 2 operand of memref type. The 2 memref types shall have the same element type.  The 2 operand should actually refer to the same contiguous memory  (like in C 2 pointers can be subtracted only if they point to the same array ... or one pass the end). 

It might be easier to carry that operator around until lowering, but I would be reluctant to add an operator to a dialect that is not "ours" (or polygeist).

**Comment by victor-eds** [comment]
> If I understand your comment correctly you are proposing to add a new operation to the memref dialect. The new operation would take 2 operand of memref type. The 2 memref types shall have the same element type. The 2 operand should actually refer to the same contiguous memory (like in C 2 pointers can be subtracted only if they point to the same array ... or one pass the end).

I was thinking on adding it to `polygeist` (we already have `subindex`, similar to `GEP` for memrefs)

**Comment by whitneywhtsang** [comment]
Would `memrefdiff(a, b)` be generated by the FE codegen portion? Or the pattern is recognized in a canonicalization pass?

**Comment by whitneywhtsang** [comment]
We have an offline conversation; the conclusion is limiting our changes to functional and not modify/remove these canonicalization patterns until we have a better understanding on the pipeline and have performance testing.

**Pull request merged**